src/main/scala/tech/cryptonomic/conseil/tezos/repositories/TezosRepositories.scala

```diff
@@ -0,0 +1,190 @@
+
+  /** Writes several ballot results referencing the containing block, returning the records stored */
+  def writeVotingBallots(ballots: List[Voting.Ballot], block: Block): Eff[Option[Int]]
+
+}
+
+/** Defines repository operations related to accounts stored in the chain.
+  * Parametric on the returned effect type wrapper
+  */
+trait AccountsRepository[Eff[_]] {
+
+  /** Reads the account ids in the checkpoint table, considering
+    * only those at the latest block level (highest value)
+    * @return the list of relevant ids and block references
+    */
+  def getLatestAccountsFromCheckpoint: Eff[Map[AccountId, BlockReference]]
+
+  /** Removes data from the accounts checkpoint table
+    * @param ids a selection of ids to remove. If none is specified the checkpoint is emptied
+    */
+  def cleanAccountsCheckpoint(ids: Option[Set[AccountId]] = None): Eff[Int]
+
+  /** Writes accounts with block data to a database, updating any existing.
+    * @param accountsInfo List data on the accounts and the corresponding blocks that operated on those
+    * @return the number of rows written (if available from the underlying driver)
+    */
+  def updateAccounts(accountsInfo: List[BlockTagged[Map[AccountId, Account]]]): Eff[Int]
+
+  /** Writes association of account ids and block data to define accounts that needs update
+    * @param accountIds will have block information, paired with corresponding account ids to store
+    * @return the rows written (if available form the underlying driver)
+    */
+  def writeAccountsCheckpoint(accountIds: List[(BlockHash, Int, List[AccountId])]): Eff[Option[Int]]
+
+}
+
+/** Defines repository operations related to chain delegates contracts.
+  * Parametric on the returned effect type wrapper
+  */
+trait DelegatesRepository[Eff[_]] {
+
+  /** Reads the delegate key hashes in the checkpoint table, considering
+    * only those at the latest block level (highest value)
+    * @return the list of relevant rows
+    */
+  def getLatestDelegatesFromCheckpoint: Eff[Map[PublicKeyHash, BlockReference]]
+
+  /** Removes data from the ddelegates checkpoint table
+    * @param pkhs a selection of key hashes to remove. If none is specified the checkpoint is emptied
+    */
+  def cleanDelegatesCheckpoint(pkhs: Option[Set[PublicKeyHash]] = None): Eff[Int]
+
+  /** Writes delegates with block data to a database, updating any existing.
+    * @param delegatesInfo List data on the delegates and the corresponding blocks that operated on those
+    * @return the number of rows written (if available from the underlying driver)
+    */
+  def updateDelegates(delegatesInfo: List[BlockTagged[Map[PublicKeyHash, Delegate]]]): Eff[Int]
+
+  /** Selects accounts corresponding to the given ids and copy the rows
+    * as delegated contracts, whose data structure should match at least in part.
+    */
+  def copyAccountsAsDelegateContracts(contractsIds: Set[ContractId]): Eff[Option[Int]]
```

Comment:
Shouldn't it be split into two methods? One in `AccountsRepository`?

Reply:
Like reading in one repo and writing in the other?

I'm afraid it works like so because the underlying sql is a single insert from select...?

Leaky abstraction?